Support readiness checks

mikedanese opened the pull request:
This is a pass at addressing #620, to support readiness checks of pods before being added to the endpoint pool of services. What's left is to do is write unit/integration tests and improve documentation. I'm opening this as a WIP to start discussion around the api and anything else.

mikedanese committed:
implement readiness.

mikedanese committed:
api changes to support readiness.

a-robinson commented:
cc @bgrant0607, want to take a look at this?

brendandburns commented on pkg/kubelet/probe.go:
```diff
@@ -38,6 +41,27 @@ var (
 	tcprober   = tcprobe.New()
 )
 
+func (kl *Kubelet) probeLiveness(podFullName string, podUID types.UID, status api.PodStatus, container api.Container, dockerContainer *docker.APIContainers) (probe.Status, error) {
```
given that these two are basically identical.  Can we refactor in terms of a helper function that takes a Probe as an argument, and then pass LivenessProbe or ReadinessProbe as appropriate?

mikedanese replied:
Done.

brendandburns commented:
Generally speaking the code LGTM.  I defer to @bgrant0607 and @smarterclayton  on the details of the API spec.  We might consider making readiness tied to a specific port, but I think that could (and should) be an extension for the future.

bgrant0607 commented:
Thanks. Will look at this today.

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -513,8 +523,8 @@ type PodSpec struct {
-	Phase PodPhase `json:"phase,omitempty"`
-
+	Phase        PodPhase        `json:"phase,omitempty"`
+	ServicePhase PodServicePhase `json:"servicePhase,omitempty"`
```
I would like this to look similar to NodeConditions:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1beta3/types.go#L766

"Phase" is reserved for monotonic progressions. "Conditions" may change back and forth. Also, I expect to add more conditions in the future. And, I'd like to be able to add more information in the future, such as transition time, reason, message, etc.

mikedanese replied:
Makes sense. PodCondition will start with on condition which will be PodInService.

bgrant0607 commented on pkg/master/pod_cache.go:
```diff
@@ -188,16 +188,21 @@ func (p *PodCache) computePodStatus(pod *api.Pod) (api.PodStatus, error) {
 	newStatus.HostIP = nodeStatus.HostIP
-
+	newStatus.Phase = api.PodUnknown
+	newStatus.ServicePhase = api.PodOutOfService
```
We're moving computation of pod status out of pod_cache.go, to Kubelet. #4194

bgrant0607 commented on pkg/kubelet/kubelet.go:
```diff
@@ -1438,6 +1449,12 @@ func (kl *Kubelet) GetPodStatus(podFullName string, uid types.UID) (api.PodStatu
 
+	for _, c := range manifest.Containers {
+		containerStatus := info[c.Name]
+		containerStatus.Ready = kl.readiness.IsReady(strings.TrimPrefix(containerStatus.ContainerID, "docker://"))
```
A container without a readiness probe should be declared ready if and only if it is in the running phase.

mikedanese replied:
Done.

mikedanese commented:
I've rebased, addressed the comments and refactored the api. I'll write the docs/tests, and finish this up sometime tomorrow/this weekend.

thockin commented on pkg/api/types.go:
```diff
@@ -403,6 +404,13 @@ const (
+
+// These are the valid service statuses of pods.
+const (
+	PodInService = "InService"
```
if a pod can only be InService, is this useful?  Shouldn't there be a second condition for this?

bgrant0607 replied:
I would like to report Liveness status, also, and probably other conditions later.

bgrant0607 replied:
I would like to follow the same pattern as NodeCondition:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1beta3/types.go#L793

```
type PodConditionKind string
const (
	PodLive PodConditionKind = "Live"
	PodReady PodConditionKind = "Ready"
)
```

bgrant0607 replied:
And:
```
type PodConditionStatus string
const (
	PodConditionFull    PodConditionStatus = "Full"
	PodConditionNone    PodConditionStatus = "None"
	PodConditionUnknown PodConditionStatus = "Unknown"
)
```
With pods degraded/partial states make a little less sense than for nodes, so I might be willing to use Good/Bad/Unknown or something similar. However, an example of a Degraded/Partial condition might be some containers in the pod reporting ready and some unresponsive, and/or some of the last N probes failing but not yet at the threshold to declare it not ready (if/when we add thresholds).

bgrant0607 replied:
Note that if we were to represent partial conditions, we'd need to ignore containers without probes defined rather than treat them as successful.

To be clear: I think we should put off reporting details about partial conditions, but am just thinking about likely future direction.

Another issue with respect to readiness is how to deal with multi-port services (#1802, #2585). If we remove pods that aren't ready from endpoints as in this PR, how would that affect multi-port services? I'm willing to punt on that for this PR. If the user specified a readiness probe using a named port on the container, we could map it back to a list of ports in the endpoint (so, per-port readiness, though only one port per container).

mikedanese replied:
NodeConditionStatus will be so similar to PodConditionStatus that NodeConditionStatus could be renamed to ConditionStatus and used for both cases, and eventually maybe for ContainerCondition. I don't think it would be detrimental to the api for some ConditionStatuses to be inapplicable to certain Condition types.

bgrant0607 replied:
On Sat, Feb 7, 2015 at 3:04 PM, Mike Danese <notifications@github.com>
wrote:

> NodeConditionStatus will be so similar to PodConditionStatus that
> NodeConditionStatus could be renamed to ConditionStatus and used for both
> cases, and eventually maybe for ContainerCondition.
>
Works for me.

thockin commented on pkg/api/types.go:
```diff
@@ -515,8 +524,8 @@ type PodSpec struct {
-	Phase PodPhase `json:"phase,omitempty"`
-
+	Phase      PodPhase       `json:"phase,omitempty"`
+	Conditions []PodCondition `json:"Condition,omitempty"`
```
It's not clear to me why this is an array?

mikedanese replied:
As per @bgrant0607's suggestion, this is intended to be like [NodCondition](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1beta3/types.go#L768-L812) as he expects to add more conditions in the future. I stopped short of creating the PodCondition struct but I could in order to fully mirror NodeCondition api. Until more PodConditions are added, Ready bool would be a functional substitute for Conditions. So I guess it comes down to whether we want to add PodCondition to the api now or hold off until later.

thockin replied:
OK

bgrant0607 replied:
I would like to add PodCondition now. I'm ok we initially omit some fields, such as LastTransitionTime, Reason, and Message.

thockin commented on pkg/api/types.go:
```diff
@@ -403,6 +404,13 @@ const (
 	PodUnknown PodPhase = "Unknown"
 )
 
+type PodCondition string
```
We use to call Phase Condition, but we changed it (I forget why).  Do we want to use the same name, or something less vague?

If this is the InServiceStatus, could we call it that?

thockin replied:
Or should this just be a bool Ready?

bgrant0607 replied:
My previous comment doesn't show here any more: I want this to look like NodeCondition.

thockin commented on pkg/kubelet/kubelet.go:
```diff
@@ -1047,8 +1051,15 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
+				}
 				// TODO: This should probably be separated out into a separate goroutine.
-				healthy, err := kl.probeLiveness(podFullName, uid, podStatus, container, dockerContainer)
+				healthy, err := kl.probeContainer(container.LivenessProbe, podFullName, uid, podStatus, container, dockerContainer, probe.Success)
```
shouldn't we probe liveness first?  If not, maybe a comment about why?

bgrant0607 replied:
I would think so.

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -431,6 +439,7 @@ type ContainerState struct {
 }
 
 type ContainerStatus struct {
+	Ready bool `json:"ready"`
```
I'm ok with this for now, though I imagine we'll need more info later, as in PodCondition or ContainerStateTerminated. However, organizationally, please put it below State.

bgrant0607 commented on pkg/api/v1beta1/types.go:
```diff
@@ -397,6 +405,7 @@ type ContainerState struct {
 }
 
 type ContainerStatus struct {
+	Ready bool `json:"ready" description:"specifies wheter the container has passed it's readiness probe"`
```
Typos: "wheter" should be "whether", and "it's" should be "its"

bgrant0607 commented on pkg/kubelet/probe.go:
```diff
@@ -39,13 +42,37 @@ var (
+	dockerContainer *docker.APIContainers,
+	defaultResult probe.Result) (result probe.Result, err error) {
+	if p == nil {
+		return probe.Success, nil
```
Why does this return Success rather than the defaultResult? In what scenario would defaultResult be failure?

bgrant0607 replied:
Ok, nevermind. A comment would be useful.

bgrant0607 commented on pkg/kubelet/probe.go:
```diff
@@ -39,13 +42,37 @@ var (
+	for i := 0; i < maxRetries; i++ {
+		result, err = kl.runProbe(p, podFullName, podUID, status, container)
+		if result == probe.Success {
+			return
```
return what?

mikedanese committed:
refactor pkg/probe ProbeRunners to interfaces and move global probers into kubelet field for testability

mikedanese committed:
added tests

mikedanese commented:
I've refactored the api per the comments, and added documentation. I've implemented readiness. I've made some minor refactoring's to pkg/probe related code to facilitate testing, and added unit tests. This is ready for review. I've tried to organize the commits in logical chunks to hopefully make review easier because of the size.

bgrant0607 commented on pkg/api/v1beta1/types.go:
```diff
@@ -412,6 +414,33 @@ type ContainerStatus struct {
+	PodReady PodConditionKind = "Ready"
+)
+
+type PodConditionStatus string
```
We should just update v1beta1 and v1beta2 to use ConditionStatus. As long as we don't change the string contents or field names, the type names shouldn't matter. They don't even get exposed via the swagger API at the moment.

mikedanese replied:
Okay will do. I was wondering if that qualified as an API break this weekend. I'll make the changes when I rebase over #4241.

bgrant0607 replied:
FYI, #4241 has been merged.

bgrant0607 commented on pkg/api/types.go:
```diff
@@ -759,24 +789,12 @@ const (
-	Reason             string              `json:"reason,omitempty"`
-	Message            string              `json:"message,omitempty"`
+	Kind               NodeConditionKind `json:"kind"`
+	Status             ConditionStatus   `json:"status"`
```
Thanks for doing this refactoring.

FYI, I'm going to try to merge #4241 first, since we're having a bunch of problems in that area right now.

bgrant0607 commented on pkg/kubelet/kubelet.go:
```diff
@@ -1048,7 +1055,16 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
 				// TODO: This should probably be separated out into a separate goroutine.
-				healthy, err := kl.probeLiveness(podFullName, uid, podStatus, container, dockerContainer)
+				ready := probe.Unknown
+				healthy, err := kl.probeContainer(container.LivenessProbe, podFullName, uid, podStatus, container, dockerContainer, probe.Success)
```
I would like people to get in the habit of using "live" rather than "healthy", precisely to clarify the distinction between live and ready. Could you please change the name of the variable to "live" here?

bgrant0607 commented on pkg/kubelet/kubelet.go:
```diff
@@ -1048,7 +1055,16 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
+				ready := probe.Unknown
+				healthy, err := kl.probeContainer(container.LivenessProbe, podFullName, uid, podStatus, container, dockerContainer, probe.Success)
+				if healthy == probe.Success {
+					ready, _ = kl.probeContainer(container.ReadinessProbe, podFullName, uid, podStatus, container, dockerContainer, probe.Failure)
```
A comment about what this does in corner cases would be helpful, such as when no readiness probe is provided, or before the initial period passes.

bgrant0607 commented on pkg/kubelet/kubelet.go:
```diff
@@ -1487,6 +1503,22 @@ func getPhase(spec *api.PodSpec, info api.PodInfo) api.PodPhase {
 	}
 }
 
+func isPodReady(spec *api.PodSpec, info api.PodInfo) bool {
```
Again, a comment about the behavior would be useful: pod is not ready if any container is not ready.

bgrant0607 commented on pkg/kubelet/kubelet.go:
```diff
@@ -1499,8 +1531,20 @@ func (kl *Kubelet) GetPodStatus(podFullName string, uid types.UID) (api.PodStatu
+
 	var podStatus api.PodStatus
 	podStatus.Phase = getPhase(&spec, info)
+	if isPodReady(&spec, info) {
```
The not-ready condition should already be added at this point?

mikedanese replied:
Should there be a not-ready condition or should the ready condition just be omitted from the pod's Conditions slice? Is the not ready condition a ConditionKind of Ready and a ConditionStatus of None or does it need a separate ConditionKind?

mikedanese replied:
I'm choosing to set the condition to:
```
[]api.PodCondition{{
	Kind:   api.PodReady,
	Status: api.ConditionNone,
}}
```
if any containers in the pod are not ready.

bgrant0607 commented on pkg/master/pod_cache.go:
```diff
@@ -192,6 +192,7 @@ func (p *PodCache) computePodStatus(pod *api.Pod) (api.PodStatus, error) {
 	} else {
 		newStatus.Info = result.Status.Info
 		newStatus.PodIP = result.Status.PodIP
+		newStatus.Conditions = result.Status.Conditions
```
Sigh. I hate the pod cache. I don't have time to check right now, but it's worth looking about what will happen if kubelet can't be contacted.

mikedanese replied:
What is the desired behavior? Should the pods be removed from load balancing? I would expect that the pod's condition should become unknown in this case.

mikedanese replied:
I'm choosing to set the conditions to:
```
[]api.PodCondition{{
	Kind:   api.PodReady,
	Status: api.ConditionUnknown,
}}
```
and remove the pod from load balancing when the node cannot be contacted.

bgrant0607 replied:
See #4188. We'd like the last known status to be returned to the user. Once we add LastTransitionTime and LastProbeTime, similar to NodeCondition, clients will be able to interpret that. PodUnknown is far less useful and more disruptive. Also, we don't want application availability to necessarily be coupled to Kubelet availability, so blacking out the application when Kubelet doesn't respond isn't the right thing. We will, however, need to improve the signals with respect to node reachability.

What I expect readiness to help the most with respect to load balancing is planned outages. I also expect to use it in numerous disruptive automation processes, such as rolling updates, auto-scaling, node draining, re-scheduling, etc.

mikedanese replied:
Ya, that makes sense. Best to just coast when statuses are unreliable. Done.

bgrant0607 replied:
Just move this newStatus.Conditions assignment to the else clause, to be symmetric with the if clause.

bgrant0607 commented:
Looks pretty good. Thanks! Just a few minor comments, and we're waiting on the other PR I mentioned.

bgrant0607 commented on pkg/kubelet/kubelet.go:
```diff
@@ -1487,6 +1511,31 @@ func getPhase(spec *api.PodSpec, info api.PodInfo) api.PodPhase {
 	}
 }
 
+// getPodReadyCondition returns ready condition if all containers in a pod are ready, else it returns an empty []api.PodCondition.
```
Please update the comment

bgrant0607 commented:
Looks good. Only 2 minor comments. Thanks for all the iterations on this.

Have you run the e2e test?

mikedanese commented:
No worries. I'm not set up with gcp yet. I'll try to get that going this morning.

mikedanese commented:
That wasn't bad.
```
Ran 20 of 20 Specs in 1388.811 seconds
SUCCESS! -- 20 Passed | 0 Failed | 0 Pending | 0 Skipped I0210 08:39:49.841956  105690 driver.go:83] All tests pass
```

bgrant0607 commented:
Great! Thanks much.

LGTM
